kubeadm: deprecate the `ClusterStatus` dependency

ereslibre opened the pull request:
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:
While `ClusterStatus` will be maintained and uploaded, it won't be
used by the internal `kubeadm` logic in order to determine the etcd
endpoints anymore.

The only exception is during the first upgrade cycle (`kubeadm upgrade
apply`, `kubeadm upgrade node`), in which we will fallback to the
ClusterStatus to let the upgrade path add the required annotations to
the newly created static pods.

**Which issue(s) this PR fixes**:
Implements https://github.com/kubernetes/enhancements/pull/1380

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:
```release-note
kubeadm: The ClusterStatus struct present in the kubeadm-config ConfigMap is deprecated and will be removed on a future version. It is going to be maintained by kubeadm until it gets removed. The same information can be found on `etcd` and `kube-apiserver` pod annotations, `kubeadm.kubernetes.io/etcd.advertise-client-urls` and `kubeadm.kubernetes.io/kube-apiserver.advertise-address.endpoint` respectively.
```

@kubernetes/sig-cluster-lifecycle @kubernetes/sig-cluster-lifecycle-pr-reviews

ereslibre committed:
kubeadm: remove `ClusterStatus` dependency

While `ClusterStatus` will be maintained and uploaded, it won't be
used by the internal `kubeadm` logic in order to determine the etcd
endpoints anymore.

The only exception is during the first upgrade cycle (`kubeadm upgrade
apply`, `kubeadm upgrade node`), in which we will fallback to the
ClusterStatus to let the upgrade path add the required annotations to
the newly created static pods.

ereslibre commented:
/assign @fabriziopandini @neolit123 @rosti

neolit123 commented:
applying hold for review.
/hold

rosti commented on cmd/kubeadm/app/util/config/cluster_test.go:
```diff
@@ -374,80 +354,84 @@ func TestGetNodeRegistration(t *testing.T) {
 				},
 			},
-			expectedError: true,
+			expectedErr: true,
```
We need to verify that correct port numbers are used. Hence we should check that 0, negative port numbers and ones above 65535 are not successfully parsed.

rosti commented on cmd/kubeadm/app/util/config/cluster.go:
```diff
@@ -180,25 +183,72 @@ func getNodeNameFromKubeletConfig(kubeconfigDir string) (string, error) {
-	// gets the ClusterStatus from kubeadm-config
-	clusterStatus, err := UnmarshalClusterStatus(data)
+func getAPIEndpoint(client clientset.Interface, nodeName string, apiEndpoint *kubeadmapi.APIEndpoint) error {
+	apiEndpoints, err := GetAllAPIEndpoints(client)
```
Ideally, we don't need to fetch all endpoints. You can just add a field selector for `spec.nodeName=<nodeName>` to the query. That will get you a list, which should contain only a single pod object.
We don't seem to be using `GetAllAPIEndpoints` anywhere else so we can certainly go without it.

ereslibre replied:
Thought that this could come handy at some point, after all since we already did one request, getting more information with a single request is "free", except for some local cycles composing the information.

ereslibre replied:
Will update nevertheless.

rosti commented on cmd/kubeadm/app/util/config/cluster.go:
```diff
@@ -180,25 +183,72 @@ func getNodeNameFromKubeletConfig(kubeconfigDir string) (string, error) {
+//
+// TODO: remove after 1.20, when the ClusterStatus struct is removed from the kubeadm-config ConfigMap;
+//       callers of this function should call getAPIEndpoint directly when that happens.
+func getAPIEndpointOrMigrate(client clientset.Interface, nodeName string, apiEndpoint *kubeadmapi.APIEndpoint) error {
```
This name is a bit misleading. We don't do any migration here.
I would suggest to keep the `getAPIEndpoint` name here. Then make a couple of helpers that have an implementation specific name in their name - `getAPIEndpointFromPod` and `getAPIEndpointFromClusterStatus`. The top level func (`getAPIEndpoint`) would call just the helpers then.

ereslibre replied:
Yes, you're right, I also found the name a bit funky. Will update.

rosti commented on cmd/kubeadm/app/util/etcd/etcd.go:
```diff
@@ -127,6 +124,104 @@ func NewFromCluster(client clientset.Interface, certificatesDir string) (*Client
+}
+
+// retrieveEtcdEndpointsFromCluster returns the list of endpoints as reported on etcd's pods labels.
+func retrieveEtcdEndpointsFromCluster(client clientset.Interface) ([]string, error) {
```
The name sounds confusing. Better clarify it as `FromPod` than just `FromCluster`.

rosti commented on cmd/kubeadm/app/util/etcd/etcd.go:
```diff
@@ -127,6 +124,104 @@ func NewFromCluster(client clientset.Interface, certificatesDir string) (*Client
+// TODO: remove after 1.20, when the ClusterStatus struct is removed from the kubeadm-config ConfigMap;
+//       callers of this function should call retrieveEtcdEndpointsFromClusterWithRetry directly when that
+//       happens.
+func retrieveEtcdEndpointsFromClusterWithRetryOrMigrate(client clientset.Interface) ([]string, error) {
```
Again, strictly speaking, this is not a migration.

rosti reviewed (commented):
Thanks @ereslibre !
Overall, I like how this is going.

ereslibre commented on cmd/kubeadm/app/apis/kubeadm/apiendpoint.go:
```diff
@@ -0,0 +1,44 @@
+)
+
+// APIEndpointFromString returns an APIEndpoint struct based on a "host:port" raw string.
+func APIEndpointFromString(apiEndpoint string) (APIEndpoint, error) {
```
I was thinking that some unit tests closer to this logic would be nice. It's indirectly tested by wider tests on this same PR, but thought some units here wouldn't harm, WDYT?

ereslibre commented:
Thanks @rosti for your review!, this is ready for another pass.

rosti commented on cmd/kubeadm/app/util/config/cluster.go:
```diff
@@ -180,23 +182,61 @@ func getNodeNameFromKubeletConfig(kubeconfigDir string) (string, error) {
+	if len(podList.Items) == 0 {
+		return "", errors.Errorf("could not find API server pod for node name %q", nodeName)
+	}
+	return podList.Items[0].Annotations[constants.KubeAPIServerAdvertiseAddressEndpointAnnotationKey], nil
```
This is a certain crash here if we find an API server pod, that was created by an old version of kubeadm and does not have the annotation. We must check for the annotation key presence. This would have been caught with the master upgrade job after merge. The reason it's not in the PR, is that there are no unit tests to these funcs.

Another thing here. Everybody with an admin account here can intentionally or accidentally create a pod that has the required labels and is on the same node. Therefore, we should probably error out if more than one pod is found on the returned list.

ereslibre replied:
> This is a certain crash here if we find an API server pod, that was created by an old version of kubeadm and does not have the annotation. We must check for the annotation key presence. This would have been caught with the master upgrade job after merge. The reason it's not in the PR, is that there are no unit tests to these funcs.

Yes, this is right, will update.

> Another thing here. Everybody with an admin account here can intentionally or accidentally create a pod that has the required labels and is on the same node. Therefore, we should probably error out if more than one pod is found on the returned list.

On the case of more than one pod I was thinking in iterating over the N pods and only error out if we get different annotations (if we have N pods, but all have the same annotation I would not error out). But thought that as a first approach the simplest solution would be better. I can update the PR.

rosti reviewed (commented):
Thanks @ereslibre !
I feel that we need to focus our unit tests on the utility funcs. That way we can make more thorough and easy to follow spec as some problems may be hidden by our broad spec tests.

ereslibre commented on cmd/kubeadm/app/util/config/cluster.go:
```diff
@@ -180,23 +183,84 @@ func getNodeNameFromKubeletConfig(kubeconfigDir string) (string, error) {
+	}
+	if len(rawEndpoint) == 0 {
+		return "", errors.Errorf("could not find API server endpoint for node name %q", nodeName)
 	}
```
@rosti: updated with this strategy, let me know what you think.

I'll add some tests for the utility functions as suggested in the meantime. Thanks!

rosti replied:
I like your idea, but I was thinking on something simpler. How about just return appropriate errors if the query list does not have a single entry? Then, error out if that single entry does not have annotation.
Let's keep this simple for a start. We can make it more complex if users have problems with it. But for now, it's better to keep it simple as we'll have to support it too.

ereslibre replied:
> I like your idea, but I was thinking on something simpler. How about just return appropriate errors if the query list does not have a single entry? Then, error out if that single entry does not have annotation.

I am fine doing this, I was only worried because I haven't experimented enough with pod mirroring and what cases we can expect. Maybe this approach is too strict, but maybe is good enough, I just don't know yet; so yes, sounds fine. I will do this.

neolit123 commented on cmd/kubeadm/app/apis/kubeadm/apiendpoint_test.go:
```diff
@@ -0,0 +1,52 @@
+				t.Errorf("no error was expected, but got %v", err)
+			} else if rt.expectedErr && err == nil {
+				t.Error("no error was found, but an error was expected")
+			}
```
we use this in other places:
```
if (err != nil) != rt.expectedErr {
	t.Error("expected error %v, got %v", error: %v",  rt.expectedErr, err != nil, err)
}
```

neolit123 commented on cmd/kubeadm/app/constants/constants.go:
```diff
@@ -370,6 +370,16 @@ const (
 	KubeControllerManagerPort = 10257
 
+	// EtcdAdvertiseClientUrlsAnnotationKey is the annotation key on every etcd pod, describing the
+	// advertise client URL
```
`URLs` plural?

neolit123 replied:
or rather the key value should be `-...client-url`

neolit123 replied:
and if so we have to adjust this in the release note:

> kubeadm: The ClusterStatus struct present in the kubeadm-config ConfigMap <snip> `kubeadm.kubernetes.io/etcd.advertise-client-urls` ...

fabriziopandini replied:
Being the etcd flag `--advertise-client-urls` I'm ok with `EtcdAdvertiseClientUrlsAnnotationKey`

neolit123 commented on cmd/kubeadm/app/util/etcd/etcd.go:
```diff
@@ -127,6 +124,69 @@ func NewFromCluster(client clientset.Interface, certificatesDir string) (*Client
+func retrieveEtcdEndpointsFromPodAnnotationWithRetry(client clientset.Interface) ([]string, error) {
+	etcdEndpoints := []string{}
+	var lastErr error
+	err := wait.ExponentialBackoff(clientsetretry.DefaultRetry, func() (bool, error) {
```
the default backoff and retry are very short.
https://github.com/kubernetes/kubernetes/blob/8d247629c38228ac18513612024c527f58ed9c57/staging/src/k8s.io/client-go/util/retry/util.go#L26-L43

i think we should use `etcdBackoff` defined in this file for consistency even if this is retrying on client-go and the rest of the usages are for the etcd client and even if it's rather long.

ereslibre replied:
Thank you, I'll double check. I simply don't know what a reasonable default is right now. Note that we will forcefully have to timeout during the first `kubeadm upgrade apply`. In the case of this `etcd`, only the `apply` is affected, subsequent `kubeadm upgrade node` will find the first etcd pod after that.

neolit123 replied:
i don't remember us discussing this on the KEP. maybe we did.
wondering what would the best way be to handle this transitional upgrade.

maybe it makes sense to only use ClusterStatus for `upgrade apply` and not fetching the annotation for the 1.18 release.

this would require etcd.NewFromCluster() accepting another argument.

i did not think much about this.

ereslibre replied:
I'll think this one through a bit.

neolit123 commented on cmd/kubeadm/app/util/config/cluster.go:
```diff
@@ -180,23 +183,69 @@ func getNodeNameFromKubeletConfig(kubeconfigDir string) (string, error) {
+	} else {
+		errs = append(errs, errors.WithMessagef(err, "could not retrieve API endpoints for node %q using cluster status", nodeName))
+	}
+	return errorsutil.NewAggregate(errs)
```
can be:
```
	var errs []error
	if err := getAPIEndpointFromPodAnnotationWithRetry(client, nodeName, apiEndpoint); err == nil {
		return nil
	}
	errs = append(errs, errors.WithMessagef(err, "could not retrieve API endpoints for node %q using pod annotations", nodeName))
	if err := getAPIEndpointFromClusterStatus(client, nodeName, apiEndpoint); err == nil {
		return nil
	}
	errs = append(errs, errors.WithMessagef(err, "could not retrieve API endpoints for node %q using cluster status", nodeName))
	return errorsutil.NewAggregate(errs)
```

ereslibre replied:
Hm, I think in your suggestion `err` is out of scope.

neolit123 replied:
seems like so.

alternatively:

```
	var err error
	var errs []error
	if err = getAPIEndpointFromPodAnnotationWithRetry(client, nodeName, apiEndpoint); err == nil {
		return nil
	}
	errs = append(errs, errors.WithMessagef(err, "could not retrieve API endpoints for node %q using pod annotations", nodeName))
	if err = getAPIEndpointFromClusterStatus(client, nodeName, apiEndpoint); err == nil {
		return nil
	}
	errs = append(errs, errors.WithMessagef(err, "could not retrieve API endpoints for node %q using cluster status", nodeName))
	return errorsutil.NewAggregate(errs)

ereslibre replied:
I think I favour the explicit `else` branches, is it too bad for you?

neolit123 replied:
yes, it's fine.
i remember staticcheck complaining about similar branching, but hopefully the scoped `err` will not trigger this.

neolit123 commented on cmd/kubeadm/app/util/etcd/etcd.go:
```diff
@@ -127,6 +124,69 @@ func NewFromCluster(client clientset.Interface, certificatesDir string) (*Client
+	return etcdEndpoints, nil
+}
+
+func retrieveEtcdEndpointsFromPodAnnotation(client clientset.Interface) ([]string, error) {
```
we can add the following bellow this line?
```
klog.V(3).Infof("retrieving the annotation %q from etcd Pods", constants.EtcdAdvertiseClientUrlsAnnotationKey)
```

fabriziopandini replied:
`getRawEtcdEndpointsFromPodAnnotation`for consistency with the API endpoint part

neolit123 commented on cmd/kubeadm/app/util/etcd/etcd.go:
```diff
@@ -127,6 +124,69 @@ func NewFromCluster(client clientset.Interface, certificatesDir string) (*Client
+	for _, pod := range podList.Items {
+		etcdEndpoint, ok := pod.ObjectMeta.Annotations[constants.EtcdAdvertiseClientUrlsAnnotationKey]
+		if !ok {
+			return []string{}, errors.Errorf("etcd pod %s is missing %s annotation; cannot infer etcd advertise client url", pod.ObjectMeta.Name, constants.EtcdAdvertiseClientUrlsAnnotationKey)
```
suggesting minor change to the message:
```
etcd Pod %q is missing the %q annotation; cannot infer etcd advertise client URL
```

fabriziopandini commented on cmd/kubeadm/app/util/etcd/etcd.go:
```diff
@@ -127,6 +124,69 @@ func NewFromCluster(client clientset.Interface, certificatesDir string) (*Client
+}
+
+// retrieveEtcdEndpointsFromPodAnnotation returns the list of endpoints as reported on etcd's pod annotations
+func retrieveEtcdEndpointsFromPodAnnotationWithRetry(client clientset.Interface) ([]string, error) {
```
same as above (drop `WithRetry`) + explain why we are waiting in the code; in this case, the main reason is to wait for static pods to get registered, is it right?

ereslibre replied:
Yes, in both cases the same reasons apply, protection against transient failures (coming from the apiserver itself, or a load balancer), and to wait for static pod mirroring.

fabriziopandini commented on cmd/kubeadm/app/util/etcd/etcd.go:
```diff
@@ -127,6 +124,69 @@ func NewFromCluster(client clientset.Interface, certificatesDir string) (*Client
 }
 
+// retrieveEtcdEndpoints returns the list of etcd endpoints.
+func retrieveEtcdEndpoints(client clientset.Interface) ([]string, error) {
```
Is there a reason to use `retrieve` instead of `get` like eg for `getAPIEndpoint`?

ereslibre replied:
No, I'll change that.

fabriziopandini reviewed (commented):
@ereslibre thanks for picking up this work!
Overall, it looks great, only some small nits from my side. I will do another pass as soon as the missing tests are in.

ereslibre commented:
Thank you for your reviews @rosti, @neolit123 and @fabriziopandini. Taking this again. Will address all your comments and ping you back when I have all tests ready as well.

ereslibre commented on cmd/kubeadm/app/util/config/cluster.go:
```diff
@@ -180,23 +183,69 @@ func getNodeNameFromKubeletConfig(kubeconfigDir string) (string, error) {
+	} else {
+		errs = append(errs, errors.WithMessagef(err, "could not retrieve API endpoints for node %q using pod annotations", nodeName))
+	}
+	if err := getAPIEndpointFromClusterStatus(client, nodeName, apiEndpoint); err == nil {
```
I added a note as well (not yet pushed), but in any case, the most visible TODO's will cascade the rest of the removals because everything will fall apart when removing the annotated ones :)

ereslibre committed:
kubeadm: optimize the upgrade path from ClusterStatus to annotations

When doing the very first upgrade from a cluster that contains the
source of truth in the ClusterStatus struct, the new kubeadm logic
will try to retrieve this information from annotations.

This changeset adds to both etcd and apiserver endpoint retrieval the
special case in which they won't retry if we are in such cases. The
logic will retry if we find any unknown error, but will not retry in
the following cases:

- etcd annotations do not contain etcd endpoints, but the overall list
  of etcd pods is greater than 0. This means that we listed at least
  one etcd pod, but they are missing the annotation.

- API server annotation is not found on the api server pod for a given
  node name, but no errors aside from that one were found. This means
  that the API server pod is present, but is missing the annotation.

In both cases there is no point in retrying, and so, this speeds up the
upgrade path when coming from a previous existing cluster.

ereslibre commented:
@rosti, @neolit123 and @fabriziopandini: this is ready for another pass of reviews. Tests are completed.

Also, I added an optimization when coming from a previous cluster without annotations, so we don't retry in that case: this logic is separated in another commit (https://github.com/kubernetes/kubernetes/pull/87656/commits/42886ae41c07dfc8acf6ccb7ad2841380e3892d8), that explains the whole logic.

Thank you!

ereslibre commented:
/retest

ereslibre commented:
Alright, everything is red because the context in client-go got merged :sweat_smile:

neolit123 commented on cmd/kubeadm/app/apis/kubeadm/apiendpoint_test.go:
```diff
@@ -0,0 +1,50 @@
+				t.Errorf("expected error %v, got %v, error: %v", rt.expectedErr, err != nil, err)
+			}
+			if !reflect.DeepEqual(apiEndpoint, rt.expectedEndpoint) {
+				t.Errorf("got api endpoint: %v; expected api endpoint: %v", apiEndpoint, rt.expectedEndpoint)
```
`expected API endpoint: %v, got: %v`

neolit123 commented on cmd/kubeadm/app/phases/etcd/local.go:
```diff
@@ -181,18 +181,23 @@ func GetEtcdPodSpec(cfg *kubeadmapi.ClusterConfiguration, endpoint *kubeadmapi.A
-			staticpodutil.NewVolumeMount(etcdVolumeName, cfg.Etcd.Local.DataDir, false),
-			staticpodutil.NewVolumeMount(certsVolumeName, cfg.CertificatesDir+"/etcd", false),
+	return staticpodutil.ComponentPod(
+		v1.Container{
```
the larger diff here could have been avoided if `v1.Container` remained on the previous line.

ereslibre replied:
If I move the `v1.Container` one line up, the return statement looks weird with the last argument `map[string]string{kubeadmconstants.EtcdAdvertiseClientUrlsAnnotationKey: etcdutil.GetClientURL(endpoint)},`

neolit123 commented on cmd/kubeadm/app/util/config/cluster.go:
```diff
@@ -181,23 +183,83 @@ func getNodeNameFromKubeletConfig(kubeconfigDir string) (string, error) {
+	var rawAPIEndpoint string
+	var lastErr error
+	// Let's tolerate some unexpected transient failures from the API server or load balancers. Also, if
+	// static pods were not yet mirrored into the API server we want to wait for this propagation.
```
from my experiments the mirroring is usually done right away once a static is running.
what cases are we trying to cover with a back of 30+ seconds?

can this backoff be reduced or made exponential?

ereslibre replied:
On some instances I've seen seconds since a `kubeadm init` finished until the static pods replicate on the apiserver.

There are two different things to consider here: one is the api endpoints, the other is the etcd endpoints.

* For API endpoints I agree that having a big wait is not a big advantage, because this code path is mainly exercised during upgrades, and in that case, pods have been running for quite some time.

* For etcd endpoints, this strategy mainly covers the `kubeadm init` -> first `kubeadm join` of a control plane. In this case, we want to wait for the static pod mirroring so the secondary control plane can discover at least one etcd endpoint. For other secondary control planes (like a third), this is not so relevant, because the first etcd static pod mirror will be present already, and we'll use etcd itself to sync the client (so it's not that important if the two existing etcd endpoints are mirrored -- with one is enough).

As for the timings I truly don't have a good estimate. I believe we can optimize later on.

neolit123 commented on cmd/kubeadm/app/util/config/cluster.go:
```diff
@@ -181,23 +183,83 @@ func getNodeNameFromKubeletConfig(kubeconfigDir string) (string, error) {
-	e, ok := clusterStatus.APIEndpoints[nodeName]
-	if !ok {
-		return errors.New("failed to get APIEndpoint information for this node")
+func getRawAPIEndpointFromPodAnnotationWithoutRetry(client clientset.Interface, nodeName string) (string, error) {
```
do we need the `...WithoutRetry` here, given there is no other `getRawAPIEndpointFromPodAnnotation` variants?
i think i saw comments on this topic before.

ereslibre replied:
The way functions were splitted were to ease unit testing for each one:

* `getAPIEndpoint` just calls to `getAPIEndpointWithBackoff` with `constants.StaticPodMirroringDefaultRetry`: no explicit testing needed here.

* `getAPIEndpointWithBackoff`: first tries to retrieve the endpoint with `getAPIEndpointFromPodAnnotation`, if it's not present, try to use the cluster status with `getAPIEndpointFromClusterStatus`. Tests ensure that this behavior happens.

* `getAPIEndpointFromPodAnnotation`: tries to fetch the api endpoint from pod annotations with a backoff. Tests ensure that this behavior happens. Calls to `getAPIEndpointFromPodAnnotationWithoutRetry`.

* `getAPIEndpointFromPodAnnotationWithoutRetry`: tries to fetch the api endpoint from pod annotations without any kind of retry. Tests ensure that this behavior happens.

* `getAPIEndpointFromClusterStatus`: already existed.

Please, note that this will be very simplified when we remove the ClusterStatus. I tried to separate all concerns on this interim while we must fallback to the cluster status if pod annotations are missing.

rosti replied:
I see where you are going, but can't we merge `getAPIEndpointWithBackoff` and `getAPIEndpoint`?

neolit123 replied:
in the code base we already have the "WithRetry" naming, but not "WithoutRetry"
it seems a bit odd that we now have a default that always retries.

ereslibre replied:
Requested here: https://github.com/kubernetes/kubernetes/pull/87656#discussion_r373976594

rosti commented on cmd/kubeadm/app/util/etcd/etcd.go:
```diff
@@ -127,6 +122,95 @@ func NewFromCluster(client clientset.Interface, certificatesDir string) (*Client
 
+// getEtcdEndpoints returns the list of etcd endpoints.
+func getEtcdEndpoints(client clientset.Interface) ([]string, error) {
+	return getEtcdEndpointsWithBackoff(client, constants.StaticPodMirroringDefaultRetry)
```
Again, can we merge `getEtcdEndpoints` and `getEtcdEndpointsWithBackoff`?

ereslibre replied:
The reason for these functions is that `getEtcdEndpoints` and `getAPIEndpoint` don't need testing. We test their `WithBackoff` counterparts, where we can control the backoff on the unit tests, so we have faster unit test execution and controlled with the backoff required depending on the test cases we are stubbing.

rosti reviewed (approved):
Thanks @ereslibre !
Looks good to me! Only minor naming nits at this point.
I'll nevertheless hold for a review by @fabriziopandini as he is the KEP author and may spot some detail that I've missed.
/lgtm
/hold

fabriziopandini commented on cmd/kubeadm/app/util/config/cluster_test.go:
```diff
@@ -727,41 +825,275 @@ func TestGetGetClusterStatus(t *testing.T) {
+		{
+			name:     "exactly one pod with annotation",
+			nodeName: nodeName,
+			backoff:  wait.Backoff{Duration: 0, Jitter: 0, Steps: 1},
```
nit: I don't think test testing the backoff logic should be done here, this is a utility already tested elsewhere

ereslibre replied:
We are note testing the backoff logic. We are testing [this logic](https://github.com/kubernetes/kubernetes/pull/87656/files#diff-60f768874354bccc5862ca90ec601fc6R210-R228), controlling the backoff, so the wait on retries is faster while running unit tests.

neolit123 replied:
it's really odd to have unit tests with a zero backoff.

ideally, there should be a function without backoff that is unit tested. the function that wraps it should have a backoff (named `...WithRetry`) and should not be unit tested.

ereslibre replied:
Let me elaborate a little more: the idea behind this was to have fast unit test executions even when playing with retries. If we are testing how our function fallbacks, we want to mimic certain retry logic, and I'd like tests to not sleep for seconds.

Ideally, the only place where we want to control backoff on unit tests is on the functions `getAPIEndpointWithBackoff` and `getEtcdEndpointsWithBackoff`, so we cover the case in which the information is not present in an annotation (or apiserver temporarily unreachable), and so we try to fallback to the cluster status.

Does this make sense? If it does, I will update the tests, and simplify tests where backoff is not providing a real benefit (e.g. `TestGetAPIEndpointFromPodAnnotation` can get rid of the `backoff` in the test table).

neolit123 replied:
i can understand the idea to have better coverage.

do we really care to pass a custom backoff from each sub-test, or can it be a variable defined near the getAPIEndpointFromPodAnnotation() call?

ereslibre replied:
It can totally be on the code block that calls the function, instead of part of the test table. Maybe I have to keep it as part of the test table (each subtest) if I'm going to test the fallback (so the waits are reasonable when unit testing).

I'll update this PR.

fabriziopandini commented on cmd/kubeadm/app/util/etcd/etcd_test.go:
```diff
@@ -106,3 +115,285 @@ func TestGetClientURLByIP(t *testing.T) {
+	}{
+		{
+			name:    "exactly one pod with annotation",
+			backoff: wait.Backoff{Duration: 0, Jitter: 0, Steps: 1},
```
nit, same as above (do not test ExponentialBackoff)

ereslibre replied:
Same, but with [this logic](https://github.com/kubernetes/kubernetes/pull/87656/files#diff-5e4c5bba67c635568de70e0c424882c6R131-R140).

fabriziopandini reviewed (commented):
@ereslibre this is turning out well.
The only nit from my side is about avoiding to test the ExponentialBackoff behavior

neolit123 commented:
/retitle kubeadm: deprecate the `ClusterStatus` dependency

ereslibre commented:
This should be ready now @kubernetes/sig-cluster-lifecycle-pr-reviews.

rosti reviewed (approved):
Thanks @ereslibre !
/lgtm

ereslibre commented:
/retest

neolit123 commented:
/priority important-longterm

neolit123 commented:
@ereslibre bazel needs a update:
> pull-kubernetes-verify — Job failed.

ereslibre commented:
> bazel needs a update:

Ouch, updated and took the opportunity to rebase on top of latest master.

fabriziopandini reviewed (commented):
Thanks @ereslibre for addressing all the comments!
Let's keep an eye on the test grid now!
/hold cancel
/approve 
/lgtm
